### ISSUE TITLE
Integrate with Git Updater for automatic issue linking

### ISSUE THREAD

**Issue by brashrebel** [realbigplugins/edd-slack#81]
Over [here](https://github.com/sandhillsdevelopment/slack) we've got a handy repo set up which makes it so entering # plus a number within Slack automatically links to an issue. It would be way cool to have a UI for making these connections.

I think integrating with Git Updater makes sense since it will already be authenticated with a repo and we can just use that to fetch available repos.

The UI would possibly be similar to the notifications one. A repeater where rules could be setup. name the rule, choose the channel(s), select the repo and possibly make up a prefix if multiple repos are used in the same channel.

**Comment by d4mation** [comment]
I don't think this will work in this case due to the difference in servers.

For the Project you linked, that is a Node.JS App. Those are running constantly, whether someone has visited the site or not. This means it can monitor a Slack Team for people using a phrase (In this case, an Issue number) and then react accordingly.

Apache/Nginx only run PHP code when the site is hit, so it would not be able to monitor a Slack Team effectively.

Unless you're thinking something like Merge Tags where within a Notification you'd place something like `%issue_number%` and it would hot-swap that out. But any Notification Trigger I can think of within EDD or any of the other Integrations we support in EDD Slack would not be linked to an Issue Number as they are things like Purchases/License Activations/etc.

Let me know if I'm misunderstanding you. I just don't think something like this could work unfortunately.

**Comment by brashrebel** [comment]
Ok, makes sense to me. Is there any alternative approach? Can something _like_ this be done with PHP/JS? Can something be setup in Slack to run our script when strings match? Can our plugin create and manage a node.js app ever at all in any circumstance?

**Comment by d4mation** [comment]
> Can something like this be done with PHP/JS?

No, this is a limitation of PHP in general and by extension WordPress. (Notes about JS next)

> Can something be setup in Slack to run our script when strings match?

Not within Slack itself, no. Slack doesn't really offer any services like that. It just has an API where you can hit different endpoints to do whatever you would like with them. It is up to you to provide implementation and interface. EDD Slack provides the implementation and interface for API interactions with Slack (In the case of a Slack App. Webhooks are also API-related, but require basically 0 permissions). Any Slack App you install through their "App Store" or whatever they call it works just like EDD Slack at its core, they are just made to be publicly accessible and in many cases necessitate being able to constantly listen to an API rather than wait for a Webhook to hit the server explicitly.

If you mean something within EDD Slack, again this is a PHP limitation. PHP cannot constantly monitor an API as the webpage needs to be hit in order for any code to run on the server. Similarly, any JavaScript _within the browser_ also needs a web page loaded. (I specified "in the browser" as Node.JS is just server-side JavaScript)

> Can our plugin create and manage a node.js app ever at all in any circumstance?

Technically, yes. There could be a separate Node.JS app set up and they could communicate to each other using a REST API of sorts. This would require the Node.JS app be hosted on a server that is open to the public though as the EDD Site would need the ability to ping the Node.JS app and if it is stuck behind a local network with no port-forwarding or anything set up then it would not work.

---

But this really comes across to me as scope creep though. While this is really cool functionality, it has little to do with the average EDD site itself. It makes assumptions that they're selling software (A fair assumption in many cases I'll admit) but also that they're using GitHub rather than, say, BitBucket or GitLab as well as that they're comfortable setting up and hosting another server for this. 

Slack's documentation is pretty bad about not distinguishing between an actual Bot (Like the Repo you linked earlier) and a "Bot User" which is what Slash Commands and Interactive Messages in EDD Slack use. For a real Bot, you have to actually create a Bot so that they have Tokens and such to send messages. For Bot Users, you don't need those. They send messages with _your_ permissions and then pretend to be whatever they want and just get "Bot" appended after their "username". This caused me to originally think we would have needed a Node.JS app or similar to facilitate Slash Commands/Interactive Notifications back here https://github.com/realbigplugins/edd-slack/issues/7#issuecomment-263885756 and (although I guess the whole conversation was in Slack-only and has likely been lost by now since we don't have permanent logs) we didn't like the idea of distributing two software packages, especially since one required its own server basically. A lot of WordPress folk use Managed Hosting and wouldn't be able to set up a Node.JS app.

If this could be published as a PUBLIC Slack App somehow and people could just add it to their Slack Team like any other one on their "App Store" I think it may work. But that may be a hassle in its own. 

I do like the idea overall for this functionality, but I don't think it should be tied down to EDD. Especially to EDD Slack since it isn't tied to any of its Notification Triggers or reporting information about the Install/Store via Slash Commands.
